### ISSUE TITLE
Code from external data / resources

### ISSUE THREAD

**Issue by koperagen** [google/ksp#793]
Hi! I want to generate classes to read data source (data set, REST API, database) represented by some sample from file or url into a typed structure. In code, it would be a file annotation that points to this file or url, and then the preprocessor would read the data from it, infer its structure and generate the code.
```
@file:ImportDataSchema(
    name = "Repository",
    url = "...",
)
```
It would be nice to use the KSP because annotations seem to be the standard input for code generators in serialization, database, DI frameworks. Plus, annotations can be in the same file as code that uses the codegen result and you don't need to specify the package for the resulting class manually. On the other hand, SQLDelight does very similar code generation from declarations in the build script with gradle plugin. 
What do you think? Is it a valid use case for the KSP?

**Comment by ting-yuan** [comment]
Yes, this sounds within KSP's capability. I'm assuming that the preprocessor isn't based on KSP, because reading non-source data is [yet to be supported](https://github.com/google/ksp/issues/431) in KSP.

**Comment by hfhbd** [comment]
Would it also possible to add an API to create KSP code by hand? 
Use-case: I want to create Kotlin code (with kotlinpoet) but would like to provide a plugin mechanism to change this generated code before writing it to a file and compiling it with kotlinc. So I thought to provide the ksp code of this in-memory representation and apply the plugins and then write the changed code (and finally compile it).
